dace/transformation/interstate/state_replication.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2019-2024 ETH Zurich and the DaCe authors. All rights reserved.
+""" State replication transformation """
+
+from dace import data as dt, sdfg as sd
+from dace.sdfg import utils as sdutil
+from dace.sdfg.state import SDFGState
+from dace.transformation import transformation
+from copy import deepcopy
+from dace.transformation.interstate.loop_detection import DetectLoop
+from dace.properties import make_properties
+
+@make_properties
+class StateReplication(transformation.MultiStateTransformation):
+    """
+    Creates a copy of a state for each of its incoming edge. Then, redirects every edge to a different copy.
+    This results in states with only one incoming edge.
+    """
+
+    target_state = transformation.PatternNode(sd.SDFGState)
+
+    @staticmethod
+    def annotates_memlets():
+        return True
+
+    @classmethod
+    def expressions(cls):
+        return [sdutil.node_path_graph(cls.target_state)]
+
+
+    def can_be_applied(self, graph, expr_index, sdfg, permissive=False):
+        target_state: SDFGState = self.target_state
+
+        out_edges = graph.out_edges(target_state)
+        in_edges = graph.in_edges(target_state)
+
+        if len(in_edges) < 2:
+            return False
+
+        # avoid useless replications
+        if target_state.is_empty() and len(out_edges) < 2:
+            return False
+
+        # make sure this is not a loop guard
+        if len(out_edges) == 2:
```

Comment:
Application on loops results in the addition of useless states (but the SDFG is still correct). This will not get rid of the loop, meaning that `apply_transformations_repeated` will never halt.